memcached.c

```diff
@@ -4242,13 +4242,32 @@ static int enable_large_pages(void) {
     return ret;
 #elif defined(__linux__) && defined(MADV_HUGEPAGE)
     /* check if transparent hugepages is compiled into the kernel */
-    struct stat st;
-    int ret = stat("/sys/kernel/mm/transparent_hugepage/enabled", &st);
-    if (ret || !(st.st_mode & S_IFREG)) {
+    /* RH based systems possibly uses a different path */
+    static const char *mm_thp_paths[2] = {
+        "/sys/kernel/mm/transparent_hugepage/enabled",
+        "/sys/kernel/mm/redhat_transparent_hugepage/enabled",
+    };
+
+    size_t pathlen = sizeof(mm_thp_paths) / sizeof(mm_thp_paths[0]);
+    char thpb[128] = {0};
+    int pfd = -1;
+    for (int i = 0; i < pathlen; i ++) {
+        if ((pfd = open(mm_thp_paths[i], O_RDONLY)) != -1)
+            break;
+    }
+
+    if (pfd == -1) {
         fprintf(stderr, "Transparent huge pages support not detected.\n");
         fprintf(stderr, "Will use default page size.\n");
         return -1;
     }
+    ssize_t rd = read(pfd, thpb, sizeof(thpb));
+    close(pfd);
+    if (rd <= 0 || strstr(thpb, "[never]")) {
```

Comment:
Lines like this tend to set off an army of student static analyzers. Need to use a string function with bounding, or at least manually insert a null terminator at the end of thpb before running strstr against it.

Reply:
the buffer is initialized but I can change it a bit.